### ISSUE TITLE
autograd engine callbacks don't respect non-default cuda streams

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#37860]
## 🐛 Bug

Not sure if this is actually a bug, but discovered this when debugging https://github.com/pytorch/pytorch/issues/37790. It seems that callbacks added to the autograd engine with `queue_callback` (from C++) or `Variable._execution_engine.queue_callback` (Python) don't respect the current cuda stream and instead execute on the default stream. As a result any cuda kernels executed in the callback would not be placed on the non-default stream. 

## To Reproduce

Patch the test in `test_cuda` in this PR: https://github.com/pytorch/pytorch/pull/37858. This test essentially enqueues a callback into the autograd engine that checks if the current stream is equal to the default stream. In `backward()` in that test, the current stream is NOT the default stream, as expected, however, in the callback, this does not hold true.

## Expected behavior
The non-default stream should be respected. Not sure if the autograd callbacks should support this natively or if it the user's responsibility to ensure the callbacks are executed on the non-default cuda stream.

## Environment

Latest master

**Comment by albanD** [comment]
cc @mruberry that added non-default stream support.

**Comment by mruberry** [comment]
@mcarilli as the current owner ;)

I believe we originally made some autograd hooks/callbacks run on the non-default stream, but I'm not sure about this case or what users would expect to happen here (off the top of my head).

**Comment by rohan-varma** [comment]
@mruberry Intuitively I'd expect this, since if the user is using post hooks for things such as modifying the gradients, then the difference in streams could cause correctness issues in terms of the grad not being correctly modified.

Looking at the `engine.cpp` code, it seems that the pre/post hooks are called in `call_function`, which is after the cuda stream guard is set. So it seems to me that the hooks should be called under the correct cuda stream.

**Comment by mruberry** [comment]
Hm... now I wonder if this just wasn't tested properly and regressed...

**Comment by albanD** [comment]
The hooks you enqueue with the functions you mentionned above are not pre/post hooks but final callbacks (that are called after the whole backward pass is finished). You can see the invocation [here](https://github.com/pytorch/pytorch/blob/0e3a05ec002c997794bd97c34a045b5800112894/torch/csrc/autograd/engine.cpp#L497-L501).

Note that these callbacks are not associated with any particular Node/Tensor. So there isn't really any stream to associate them with.

**Comment by mcarilli** [comment]
Another issue to keep in mind is that backward internally executes on a separate thread (potentially multiple threads, one per cuda device).  Each thread maintains its own stream pool, so the default stream on the backward thread(s) have no relation to the default or current stream on the main thread (aka the Python thread that's making Pytorch calls).  I'm not sure if the queued callback invocation executes in the side thread or the main thread.

**Comment by albanD** [comment]
It executes on whichever thread executes the last Node in the graph. So it can be either a side thread or the main one.

**Comment by rohan-varma** [comment]
@albanD So based on this discussion (specifically there not being a stream to associate them with), would the code that defines the final callback be responsible for setting the correct stream? For our use case (DDP) it seems pretty straightforward to set this (we would just look at the associated tensors we are all-reducing). Wondering if it would be useful in the autograd engine to take an optional argument for the final_callback, which is the stream to set.

**Comment by albanD** [comment]
For the final_callback, the one thing that could make sense would be to make sure that the final callback runs on the same stream as the one set by the user in the main thread where the backward was called.

For DDP, you will need different stream for different Tensors. So you can simply set these on your hook when you process the different Tensors no?

**Comment by mruberry** [comment]
@mcarilli may need to take a look. We sync the streams the autograd engine uses with the default stream(s) at the end of backward. I'm not sure if that sync occurs before or after these callbacks are called. If it's before, you may not need to use different streams (unless you want to).

**Comment by albanD** [comment]
The sync happens just after (see code https://github.com/pytorch/pytorch/blob/0e3a05ec002c997794bd97c34a045b5800112894/torch/csrc/autograd/engine.cpp#L497-L513).

**Comment by mruberry** [comment]
I wonder if it should happen before?

**Comment by albanD** [comment]
I think there are three issues here:
1) Are the execution streams used in the backward properly synced with the default stream?
  - In particular, we recently changed this to run on the thread that finishes the last Node and not the thread that called the backward. So the default stream we sync with might not be the one the user expects
2) For the same reason, the final callback might run on the default stream as well.
3) Are we syncing too late for the final callback? Because the gradients should be available there and so we should have synced before calling the callbacks.
